shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -541,5 +542,120 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
     typeMethods(typeTypeDefs(ctx.copy(env = allEnv, mthEnv = allMthEnv, tyDefs = allDefs)))
   }
   
-  
+  /**
+    * Finds the variances of all type variables in the given type definitions with the given
+    * context using a fixed point computation. The algorithm starts with each type variable
+    * as bivariant and then keeps updating based on the types it encounters.
+    *
+    * @param tyDefs
+    * @param ctx
+    * @return Mapping from type variable to variance info
+    */
+  def computeVariances(tyDefs: List[TypeDef], ctx: Ctx): Unit = {
+    var varianceUpdated: Bool = false;
+
+    def fieldVarianceHelper(fieldTy: FieldType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit = {
+        fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip))
+        updateVariance(fieldTy.ub, curVariance)
+    }
+
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit =
+      trace(s"updateVariance(ty: $ty of ${ty.getClass()}, curVariance: $curVariance)(${tyDef.nme.name}.${tyDef.tvarVariances})") {
+        ty match {
+          case ProxyType(underlying) => updateVariance(underlying, curVariance)
+          case TraitTag(_) | ClassTag(_, _) => ()
+          case ExtrType(pol) => ()
+          case t: TypeVariable => val _ = tyDef.tvarVariances.updateWith(t) {
+            case None => Some(curVariance)
+            case Some(value) =>
+              val newVariance = value && curVariance
+              if (newVariance =/= curVariance) varianceUpdated = true;
+              Some(newVariance)
+          }
```

Comment:
Actually I think this logic isn't correct because in the `None` case we should definitely set `varianceUpdated = true`. Also shouldn't the test in the `Some` case be `newVariance =/= value`?!

I'd rewrite the code like this instead:
```suggestion
          case t: TypeVariable =>
            val oldVariance = tyDef.tvarVariances.getOrElseUpdate(t, VarianceInfo.co)
            val newVariance = oldVariance && curVariance
            if (newVariance =/= oldVariance) {
              tyDef.tvarVariances(t) = newVariance
              varianceUpdated = true
            }
```